### PR TITLE
Rebase of #1059 (GEOMESA-1402 Kafka DS secondary index) on master.

### PR DESCRIPTION
**Pull request by mdzimmerman**
Second try! Creates a module called geomesa-kafka-datastore-common
with common code shared by geomesa-kafka-{08,09,10}-datastore.

Signed-off-by: Matthew Zimmerman <matt.zimmerman@ccri.com>

### CODE REVIEW

File: build/cqs.tsv [hunk 1]
```diff
@@ -74,6 +76,7 @@ net.java.dev.jsr-275:jsr-275	1.0-beta-2	compile
 net.minidev:asm	1.0.2	compile
 net.minidev:json-smart	2.1.1	compile
 net.sf.geographiclib:GeographicLib-Java	1.44	compile
+org.antlr:antlr4-runtime	4.5	compile
```
Comment (jnh5y) [inline build/cqs.tsv]: @elahrvivaz, @jahhulbert-ccri @mdzimmerman good news!  We have CQs for these 3 additions.  Antlr and Concurrent trees are completely approved, and we have parallel check-in for CQEngine.

That means that we can :shipit: when ready.

**Comment by tkunicki** [comment]
i rebased this against locationtech/master, might want to check/regenerate `build/cgs.tsv` as I had to hand merge and i am not confident w/ my work

**Comment by tkunicki** [comment]
Ok, i think the success of PR #1114 shows this as an issue w/ the travis cache for this PR.  Someone with permissions with have to [clear the PR cache](https://docs.travis-ci.com/user/caching/#Clearing-Caches)

**Comment by elahrvivaz** [comment]
@mdzimmerman @tkunicki is this ready for merging?

**Comment by tkunicki** [comment]
+1 looks good to me as far as integration into new 2.10/2.11 build and source reorg

**Comment by elahrvivaz** [comment]
@jnh5y ok if I merge?

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/pom.xml [hunk 2]
```diff
@@ -119,12 +126,13 @@
...
             <groupId>org.apache.curator</groupId>
             <artifactId>curator-test</artifactId>
-            <version>2.7.1</version>
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>joda-time</groupId>
-            <artifactId>joda-time</artifactId>
+            <groupId>org.locationtech.geomesa</groupId>
+            <artifactId>geomesa-kafka-datastore-common_2.11</artifactId>
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/pom.xml]: Does this need to be _${scala.binary.version}?

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/src/main/scala/org/locationtech/geomesa/kafka08/KafkaGeoMessageEncoding.scala [hunk 3]
```diff
@@ -10,6 +10,7 @@ package org.locationtech.geomesa.kafka08
 
 import kafka.message.{Message, MessageAndMetadata}
 import kafka.producer.KeyedMessage
+import org.locationtech.geomesa.kafka._
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/src/main/scala/org/locationtech/geomesa/kafka08/KafkaGeoMessageEncoding.scala]: Is this import used?

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-09-datastore/pom.xml [hunk 4]
```diff
@@ -40,6 +40,10 @@
 
     <dependencies>
         <dependency>
+            <groupId>org.locationtech.geomesa</groupId>
+            <artifactId>geomesa-kafka-datastore-common_2.11</artifactId>
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-09-datastore/pom.xml]: Ditto re _${scala.version.whatnot}

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-datastore-common/src/main/scala/org/locationtech/geomesa/kafka/LiveFeatureCache.scala [hunk 5]
```diff
@@ -0,0 +1,23 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.kafka
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-datastore-common/src/main/scala/org/locationtech/geomesa/kafka/LiveFeatureCache.scala]: Does the code in this module depend on Kafka at all?

Comment (mdzimmerman) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-datastore-common/src/main/scala/org/locationtech/geomesa/kafka/LiveFeatureCache.scala]: No, there are no Kafka imports in ``geomesa-kafka-datastore-common``.

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/pom.xml (hunk 2 above)
Comment (mdzimmerman) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/pom.xml]: It was my (possibly mistaken) understanding that @tkunicki 's script required that internal geomesa dependencies should have a literal ``_2.11`` and that ``${scala.binary.version}`` should be used with external dependencies like Kafka, Specs2, etc. See for example the Accumulo DS pom. If I am wrong, please correct!

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/src/main/scala/org/locationtech/geomesa/kafka08/KafkaGeoMessageEncoding.scala (hunk 3 above)
Comment (mdzimmerman) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/src/main/scala/org/locationtech/geomesa/kafka08/KafkaGeoMessageEncoding.scala]: Yes, for GeoMessage, GeoMessageEncoder, Clear, Delete, and CreateOrUpdate.

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-09-datastore/pom.xml (hunk 4 above)
Comment (mdzimmerman) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-09-datastore/pom.xml]: See above.

File: geomesa-memory/src/main/java/org/locationtech/geomesa/memory/cqengine/index/GeoIndex.java [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.JavaConversions;
+import scala.runtime.AbstractFunction1;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+public class GeoIndex<A extends Geometry, O extends SimpleFeature> extends AbstractAttributeIndex<A, O> implements OnHeapTypeIndex {
```
Comment (Afterglow375) [inline geomesa-memory/src/main/java/org/locationtech/geomesa/memory/cqengine/index/GeoIndex.java]: [picky] Same issue as above. See the "Type Parameter Naming Conventions" section here: https://docs.oracle.com/javase/tutorial/java/generics/types.html

TL:DR - use T for generic types, then S, U, V, etc for additional generic types.

File: geomesa-memory/src/main/java/org/locationtech/geomesa/memory/cqengine/query/GeoToolsFilterQuery.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    protected boolean matchesNonSimpleAttribute(Attribute<SimpleFeature, SimpleFeature> attribute, SimpleFeature object, QueryOptions queryOptions) {
+        return filter.evaluate(object);
+    }
+
+    //NB: This is IDEA's auto-generated equals.
+    @Override
+    public boolean equals(Object o) {
```
Comment (Afterglow375) [inline geomesa-memory/src/main/java/org/locationtech/geomesa/memory/cqengine/query/GeoToolsFilterQuery.java]: Verify that this actually works properly?

File: geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+import org.locationtech.geomesa.utils.geotools._
+import org.opengis.feature.`type`.AttributeDescriptor
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+
+import scala.collection.JavaConversions._
+
+class GeoCQEngine(sft: SimpleFeatureType,
+                  enableFidIndex: Boolean = false) extends LazyLogging {
+  //val cqcache = CQIndexingOptions.buildIndexedCollection(sft)
```
Comment (Afterglow375) [inline geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala]: Remove commented line?

File: geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      case None =>
+    }
+    cqcache.add(sf)
+  }
+
+  // NB: We expect that FID filters have been handled previously
+  def getReaderForFilter(filter: Filter): FR =
+    filter match {
+      case f: IncludeFilter => include(f)
+      case f => queryCQ(f, dedup = true)
```
Comment (Afterglow375) [inline geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala]: [picky] Seems strange to pass in "dedup = true" as it's already a default arg. Remove this param?

File: geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+          classOf[java.lang.Integer].isAssignableFrom(c) ||
+            classOf[java.lang.Long].isAssignableFrom(c) ||
+            classOf[java.lang.Float].isAssignableFrom(c) ||
+            classOf[java.lang.Double].isAssignableFrom(c) ||
+            classOf[java.util.Date].isAssignableFrom(c) => addNavigableIndex(ad)
+          case c if classOf[java.lang.String].isAssignableFrom(c) => addRadixIndex(ad)
+          case c if classOf[Geometry].isAssignableFrom(c) => addGeoIndex(ad)
+          case c if classOf[UUID].isAssignableFrom(c) => addUniqueIndex(ad)
+          // TODO: Decide how boolean fields should be indexed
+          case c if classOf[java.lang.Boolean].isAssignableFrom(c) => addHashIndex(ad)
```
Comment (Afterglow375) [inline geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala]: Should add a default case to be safe. Also, I think you're missing some cases like for List or Map: https://github.com/locationtech/geomesa/blob/master/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala#L530

File: geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/CQEngineQueryVisitor.scala [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+  }
+
+  /**
+    * PropertyIsNotEqualTo
+    */
+  override def visit(filter: PropertyIsNotEqualTo, data: scala.Any): AnyRef = {
+    val prop = getProp(filter)
+    val attribute: Attribute[SimpleFeature, Any] = lookup.lookup[Any](prop.name)
+    val value: Any = prop.literal.evaluate(null, attribute.getAttributeType)
+    // TODO: could this be done better?
```
Comment (Afterglow375) [inline geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/CQEngineQueryVisitor.scala]: Not too sure about leaving nebulous comments like this in. Thoughts? @jahhulbert-ccri @elahrvivaz @jnh5y

File: geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/SFTAttributes.scala [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+case class SFTAttributes(sft: SimpleFeatureType) {
+  private val attributes = sft.getAttributeDescriptors
+
+  private val lookupMap: Map[String, Attribute[SimpleFeature, _]] = attributes.map { attr =>
+    val name = attr.getLocalName
+    name -> buildSimpleFeatureAttribute(attr.getType.getBinding, name)
+  }.toMap
+
+  // TODO: this is really, really bad :)
```
Comment (Afterglow375) [inline geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/SFTAttributes.scala]: Document this as something to improve in a new ticket if it's bad?

**Comment by Afterglow375** [comment]
Has this been tested performance wise? I'm curious how much better using CQEngine is than using just the guava cache.

File: geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/pom.xml (hunk 2 above)
Comment (tkunicki) [inline geomesa-kafka/geomesa-kafka-datastore/geomesa-kafka-08-datastore/pom.xml]: use of `_2.11` follows convention for geomesa artifacts as dependencies.

File: geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/SFTAttributes.scala (hunk 12 above)
Comment (mdzimmerman) [inline geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/SFTAttributes.scala]: Yeah, we probably should have deleted this comment too. :/

**Comment by mdzimmerman** [comment]
Yes. Queries against test data are about 20-40% faster in GeoServer, depending on what indexes are present.

**Pull request merged**